compiler/src/dotty/tools/dotc/quoted/QuoteContext.scala

```diff
@@ -0,0 +1,35 @@
+package dotty.tools.dotc.quoted
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.quoted.PickledQuotes
+import dotty.tools.dotc.tastyreflect.ReflectionImpl
+
+import scala.quoted.{Expr, Type}
+import scala.runtime.quoted.Unpickler._
+
+class QuoteContext private (val tasty: scala.tasty.Reflection) extends scala.quoted.QuoteContext {
+
+  def unpickleExpr[T](pickledExpr: PickledExpr, args: PickledExprArgs): given scala.quoted.QuoteContext => Expr[T] = given qctx => {
+    new scala.internal.quoted.TastyTreeExpr(
+      PickledQuotes.unpickleExpr(pickledExpr, args) given qctx.tasty.rootContext.asInstanceOf[Context]
+    ).asInstanceOf[Expr[T]]
+  }
+
+  def unpickleType[T](pickledType: PickledType, args: PickledTypeArgs): given scala.quoted.QuoteContext => Type[T] = given qctx => {
+    new scala.internal.quoted.TreeType(
+      PickledQuotes.unpickleType(pickledType, args) given qctx.tasty.rootContext.asInstanceOf[Context]
+    ).asInstanceOf[Type[T]]
+  }
```

Comment:
I feel confused with the type signature of the two methods : why the result type cannot be just `Expr[T]` or `Type[T]`, given that `this` is the type `QuteContext`?

Reply:
Yes, they should not return the contextual function type